STRUCTURE.md

```diff
@@ -0,0 +1,123 @@
+necessarily needing to understand precisely how this gets done.
+
+Both of the examples would use the C-esque `cryptography-libsodium` for
+implementations: the differences are their focus and the APIs they expose. For
+'algorithmic' category packages, the focus is on providing easy and secure use
+of an _algorithm_ in Haskell, without considering use case; for 'functional'
+category packages, the focus is on providing an easy and secure way to _do some
+task_, without focus on the algorithm or algorithms chosen to do so.
+
+These packages, unlike the C-esque layer, are expected to provide more
+abstraction, both to avoid the complexities of C and the FFI, but also to
+address various expectations that Haskellers have, regarding things such as:
+
+* No unnecessary use of `IO` (even if this is by way of 'safe' `unsafePerformIO`
+  use);
+* Implementation of standard, law-abiding type classes, such as `Eq`, in a
+  correct and secure way;
+* Wrappers around raw `ForeignPtr`s for data that is meant to be passed around,
+  or generated in a very specific way.
+
+Specifically, the following principles generally apply:
+
+* These libraries cannot depend on non-boot packages;
+* Abstraction should be kept minimal. `newtype`-wrapping a `ForeignPtr` is fine,
+  type class hierarchies without laws are not.
+* Whenever possible, if there are multiple secure ways to do something, without
+  a definite 'this one is always better' option, users of these libraries should
+  be given _all_ options, as well as a detailed explanation of why they should
+  prefer any given one.
+* Partiality must not exist unless specifically marked as unsafe, through a
+  dedicated `Unsafe` module or prefixing the name with `unsafe`. This extends to
+  instances of known partial type classes (such as `Read`).
+
+These principles apply to 'algorithmic' category packages specifically:
+
+* The packages should document as much as possible that is known about the
+  algorithm, providing relevant links for detail where it would be too hard to
+  include.
+* The typical _secure_ uses should be enumerated.
+* Any caveats to performance should be documented clearly.
+* 'The algorithm is the focus': abstractions should only be done to hide the use
+  of C FFI, or where the abstraction is both obvious and unarguable. Use cases
+  should not be pre-supposed.
+
+These principles apply to 'functional' category packages specifically:
+
+* 'The task is the focus': specifics regarding algorithms or implementations
+  should only factor into the abstractions if there really is no other way.
+* There should be one, and only one, correct way to perform the task the package
+  is designed for. If there are legitimately multiple right ways to do
+  something, their differences must be _clearly_ enumerated and marked, with a
+  focus on non-security-familiar users.
+* Doing the right thing should be easy; doing the wrong thing should be
```

Comment:
This one is incredibly ambitious :eyes: .

You can expose a gun with 50 safeties and users somehow always manage to shoot themselves in the foot. Doing the wrong thing should be hard but impossible is a challenger users almost always win at.

Reply:
It is ambitious, but having goals is important. We can't avoid _every_ footgun, but fewer footguns is better than more footguns.

Reply:
I've put a bit of nuance in the sentence but the spirit stays the same.